Metabase attempts connection to wrong host when SSH tunnel is configured

- Your browser and the version: **Firefox 58.0.1**
- Your operating system: **Windows 10**
- Your databases: **PostgreSQL**
- Metabase version: **0.27.2**
- Metabase hosting environment: **jar on Ubuntu 16.04.3**
- Metabase internal database: **H2**

With Metabase on `server1`, PostgreSQL database on `server2`. `server2` only allows outside connection via SSH on port 2234. PostgreSQL on `server2` listening on port 5432 but only allows local connections.

Database details:

- Name: **My Database**
- Host: **localhost**
- Port: **5432**
- Database name: **my_database**
- Database username: **william**
- Database password: **supersecret**
- Use an SSH-tunnel for database connections: **On**
  - SSH tunnel host: **server2**
  - SSH tunnel port: **2234**
  - SSH tunnel username: **william**
  - SSH tunnel password: **supersecret**

After saving this configuration, I see the following log message:

    02-02 14:18:12 INFO util.ssh :: creating ssh tunnel william@server2:2234 -L 43831:localhost:5432
    Feb 02, 2018 2:18:12 PM org.postgresql.Driver connect
    SEVERE: Connection error:
    org.postgresql.util.PSQLException: Connection to server2:43831 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.
            at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:265)
            at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)
            at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:194)
            at org.postgresql.Driver.makeConnection(Driver.java:450)
            at org.postgresql.Driver.connect(Driver.java:252)
            at java.sql.DriverManager.getConnection(DriverManager.java:664)
            at java.sql.DriverManager.getConnection(DriverManager.java:208)
            at clojure.java.jdbc$get_connection.invokeStatic(jdbc.clj:358)
            at clojure.java.jdbc$get_connection.invoke(jdbc.clj:225)
            at clojure.java.jdbc$db_query_with_resultset.invokeStatic(jdbc.clj:967)
            at clojure.java.jdbc$db_query_with_resultset.invoke(jdbc.clj:936)
            at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1004)
            at clojure.java.jdbc$query.invoke(jdbc.clj:973)
            at clojure.java.jdbc$query.invokeStatic(jdbc.clj:987)
            at clojure.java.jdbc$query.invoke(jdbc.clj:973)
            at metabase.driver.generic_sql$can_connect_QMARK_.invokeStatic(generic_sql.clj:225)
            at metabase.driver.generic_sql$can_connect_QMARK_.invoke(generic_sql.clj:222)
            at metabase.driver$fn__24495$G__24287__24502.invoke(driver.clj:34)
            at metabase.driver$can_connect_with_details_QMARK_$fn__24595.invoke(driver.clj:405)
            at clojure.core$binding_conveyor_fn$fn__4676.invoke(core.clj:1938)
            at clojure.lang.AFn.call(AFn.java:18)
            at java.util.concurrent.FutureTask.run(FutureTask.java:266)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
            at java.lang.Thread.run(Thread.java:748)
    Caused by: java.net.ConnectException: Connection refused (Connection refused)
            at java.net.PlainSocketImpl.socketConnect(Native Method)
            at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
            at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
            at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
            at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
            at java.net.Socket.connect(Socket.java:589)
            at org.postgresql.core.PGStream.<init>(PGStream.java:68)
            at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:144)
            ... 24 more
    02-02 14:18:12 ERROR metabase.driver :: Failed to connect to database: org.postgresql.util.PSQLException: Connection to server2:43831 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.

Metabase is attempting to connect to `server2` when it should be attempting to connect to `localhost`, so that the connection can be passed through the tunnel.

Hello there, i'm not much of a technician but here is the only place i can seem to find some info on my problem.
Trying to add a Postgres database on Metabase via SSH tunneling gives me the following error message: "Server error encountered"
My logs are very similar to @williamjacksn 's logs on the post: http://discourse.metabase.com/t/ssh-tunnel-failure/2469
I am certain that my credentials are OK since I use the same for a Postico SSH connection as well as a DBeaver connection.

I already checked the following fix: https://github.com/metabase/metabase/pull/6970
Howerver I have no idea how to use it as i cannot find the ssh.clj file on my machine.

Would there be a fix for this ?

Extra info:
Install via the .jar file. Working on Firefox 58.0.2 | Mac OSX 10.10.5

Thanks in advance

Log:
`03-12 10:16:48 INFO util.ssh :: creating ssh tunnel metabase@192.168.1.2:22 -L 56334:localhost:5432
03-12 10:16:53 ERROR metabase.driver :: Failed to connect to database: Timed out after 5000 milliseconds.
03-12 10:16:53 DEBUG metabase.middleware :: POST /api/setup/validate 400 (5 s) (0 DB calls)
{:errors {:dbname "Timed out after 5000 milliseconds."}}

Mar 12, 2018 10:16:53 AM org.postgresql.Driver connect
SEVERE: Connection error: 
org.postgresql.util.PSQLException: The connection attempt failed.
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:275)
	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)
	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:194)
	at org.postgresql.Driver.makeConnection(Driver.java:450)
	at org.postgresql.Driver.connect(Driver.java:252)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:678)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:190)
	at clojure.java.jdbc$get_connection.invokeStatic(jdbc.clj:364)
	at clojure.java.jdbc$get_connection.invoke(jdbc.clj:226)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invokeStatic(jdbc.clj:1014)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invoke(jdbc.clj:996)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1090)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1047)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1063)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1047)
	at metabase.driver.generic_sql$can_connect_QMARK_.invokeStatic(generic_sql.clj:220)
	at metabase.driver.generic_sql$can_connect_QMARK_.invoke(generic_sql.clj:217)
	at metabase.driver$fn__25577$G__25354__25584.invoke(driver.clj:51)
	at metabase.driver$can_connect_with_details_QMARK_$fn__25710.invoke(driver.clj:451)
	at clojure.core$binding_conveyor_fn$fn__4676.invoke(core.clj:1938)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.base/java.lang.Thread.run(Thread.java:844)
Caused by: java.net.SocketTimeoutException: connect timed out
	at java.base/java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.base/java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:400)
	at java.base/java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:243)
	at java.base/java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:225)
	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:402)
	at java.base/java.net.Socket.connect(Socket.java:591)
	at org.postgresql.core.PGStream.<init>(PGStream.java:68)
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:144)
	... 24 more

Mar 12, 2018 10:16:58 AM org.postgresql.Driver connect
SEVERE: Connection error: 
org.postgresql.util.PSQLException: The connection attempt failed.
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:275)
	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)
	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:194)
	at org.postgresql.Driver.makeConnection(Driver.java:450)
	at org.postgresql.Driver.connect(Driver.java:252)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:678)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:190)
	at clojure.java.jdbc$get_connection.invokeStatic(jdbc.clj:364)
	at clojure.java.jdbc$get_connection.invoke(jdbc.clj:226)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invokeStatic(jdbc.clj:1014)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invoke(jdbc.clj:996)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1090)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1047)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1063)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1047)
	at metabase.driver.generic_sql$can_connect_QMARK_.invokeStatic(generic_sql.clj:220)
	at metabase.driver.generic_sql$can_connect_QMARK_.invoke(generic_sql.clj:217)
	at metabase.driver$fn__25577$G__25354__25584.invoke(driver.clj:51)
	at metabase.driver$can_connect_with_details_QMARK_$fn__25710.invoke(driver.clj:451)
	at clojure.core$binding_conveyor_fn$fn__4676.invoke(core.clj:1938)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.base/java.lang.Thread.run(Thread.java:844)
Caused by: java.net.SocketTimeoutException: connect timed out
	at java.base/java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.base/java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:400)
	at java.base/java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:243)
	at java.base/java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:225)
	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:402)
	at java.base/java.net.Socket.connect(Socket.java:591)
	at org.postgresql.core.PGStream.<init>(PGStream.java:68)
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:144)
	... 24 more`

--- a/src/metabase/util/ssh.clj
+++ b/src/metabase/util/ssh.clj
@@ -82,7 +82,7 @@
           [connection tunnel-entrance-port] (start-ssh-tunnel (assoc details :host host)) ;; don't include L7 protocol in ssh tunnel
           details-with-tunnel (assoc details
                                 :port tunnel-entrance-port ;; This parameter is set dynamically when the connection is established
-                                :host (str proto (:tunnel-host details))
+                                :host (str proto "localhost")
                                 :tunnel-entrance-port tunnel-entrance-port ;; the input port is not known until the connection is opened
                                 :tunnel-connection connection)]
       details-with-tunnel)

